File: src/Day3.kt
import java.io.File
import java.lang.Exception

fun main() {
    val input = File("src/day3.txt").readLines()

    val compartmented = input.map {
        it.chunked(it.length / 2).map(String::toSet)
    }

    val elvesGrouped = input.chunked(3).map{it.map(String::toSet)}

    fun findBadge(group: List<Set<Char>>) : Char =
        (group[0] intersect group[1] intersect group[2]).first()

    fun findDupeItem(bag: List<Set<Char>>): Char = (bag[0] intersect bag[1]).first()

    fun letterToPriority(letter: Char): Int {
        return when (letter.code) {
            in 97..122 -> letter.code - 96
            in 65..90 -> letter.code - 38
            else -> throw Exception("Not a valid char")
        }
    }

    val answer = compartmented
            .map(::findDupeItem)
            .map(::letterToPriority)
            .sum()

    val part2 = elvesGrouped
            .map(::findBadge)
            .map(::letterToPriority)
            .sum()

    println(answer)
    print(part2)
}

File: src/Day3.py
with open('src/day3.txt') as f:
    input = f.readlines()

inputStripped = [line.strip() for line in input]

compartmented = []
for line in inputStripped:
    length = int(len(line) / 2)
    frontHalf = line[:length]
    backHalf = line[length:]
    compartmented.append((frontHalf, backHalf))

setOfBadges = []
for half in compartmented:
    front = set(half[0])
    back = set(half[1])
    setOfBadges.append(front.intersection(back))


def letterToPriority(letter):
    if ord(letter) > 97:
        return ord(letter) - 96
    else:
        return ord(letter) - 38


# Part 1
print(sum([letterToPriority(letter.pop()) for letter in setOfBadges]))

everyThree = [inputStripped[i:i+3] for i in range(0, len(inputStripped), 3)]

everyThreeSet = [list(map(set, each)) for each in everyThree]

badgeLetter = []
for elf in everyThreeSet:
    badgeLetter.append(elf[0].intersection(elf[1], elf[2]))

# Part 2
print(sum([letterToPriority(letter.pop()) for letter in badgeLetter]))
